File: use-cases/school-timetabling/src/test/java/org/acme/schooltimetabling/rest/TimeTableResourceTest.java
package org.acme.schooltimetabling.rest;

import static io.restassured.RestAssured.get;
import static io.restassured.RestAssured.given;
import static org.awaitility.Awaitility.await;
import static org.hamcrest.Matchers.empty;
import static org.hamcrest.Matchers.equalTo;
import static org.hamcrest.Matchers.is;
import static org.hamcrest.Matchers.not;
import static org.hamcrest.Matchers.nullValue;

import java.time.DayOfWeek;
import java.time.Duration;
import java.time.LocalTime;
import java.util.ArrayList;
import java.util.List;

import org.acme.schooltimetabling.domain.Lesson;
import org.acme.schooltimetabling.domain.Room;
import org.acme.schooltimetabling.domain.TimeTable;
import org.acme.schooltimetabling.domain.Timeslot;
import org.junit.jupiter.api.Test;
import ai.timefold.solver.core.api.solver.SolverStatus;

import io.quarkus.test.junit.QuarkusTest;
import io.restassured.http.ContentType;

@QuarkusTest
public class TimeTableResourceTest {

    @Test
    public void solveDemoDataUntilFeasible() {
        TimeTable testTimeTable = createTestTimeTable();

        String jobId = given()
                .contentType(ContentType.JSON)
                .body(testTimeTable)
                .expect().contentType(ContentType.TEXT)
                .when().post("/timetables")
                .then()
                .statusCode(200)
                .extract()
                .asString();

        await()
                .atMost(Duration.ofMinutes(1))
                .pollDelay(Duration.ofSeconds(5))
                .pollInterval(Duration.ofSeconds(5))
                .until(() -> SolverStatus.NOT_SOLVING.name().equals(get("/timetables/"+ jobId + "?retrieve=STATUS")
                        .body()
                        .as(TimeTable.class).getSolverStatus()));

        get("/timetables").then().assertThat()
                .body("solverStatus", equalTo(SolverStatus.NOT_SOLVING.name()))
                .body("timeslotList", is(not(empty())))
                .body("roomList", is(not(empty())))
                .body("lessonList", is(not(empty())))
                .body("lessonList.timeslot", not(nullValue()))
                .body("lessonList.room", not(nullValue()));
    }

    private static TimeTable createTestTimeTable() {
        List<Timeslot> timeslotList = new ArrayList<>(10);
        long nextTimeslotId = 0L;
        timeslotList.add(new Timeslot(nextTimeslotId++, DayOfWeek.MONDAY, LocalTime.of(8, 30), LocalTime.of(9, 30)));
        timeslotList.add(new Timeslot(nextTimeslotId++, DayOfWeek.MONDAY, LocalTime.of(9, 30), LocalTime.of(10, 30)));
        timeslotList.add(new Timeslot(nextTimeslotId++, DayOfWeek.MONDAY, LocalTime.of(10, 30), LocalTime.of(11, 30)));
        timeslotList.add(new Timeslot(nextTimeslotId++, DayOfWeek.MONDAY, LocalTime.of(13, 30), LocalTime.of(14, 30)));
        timeslotList.add(new Timeslot(nextTimeslotId++, DayOfWeek.MONDAY, LocalTime.of(14, 30), LocalTime.of(15, 30)));

        long nextRoomId = 0L;
        List<Room> roomList = List.of(new Room(nextRoomId++, "Room A"), new Room(nextRoomId++, "Room B"));

        List<Lesson> lessonList = new ArrayList<>();
        long nextLessonId = 0L;
        lessonList.add(new Lesson(nextLessonId++, "Math", "A. Turing", "9th grade"));
        lessonList.add(new Lesson(nextLessonId++, "English", "I. Jones", "9th grade"));
        lessonList.add(new Lesson(nextLessonId++, "Physics", "M. Curie", "10th grade"));
        lessonList.add(new Lesson(nextLessonId++, "Geography", "C. Darwin", "10th grade"));

        return new TimeTable("test", timeslotList, roomList, lessonList);
    }
}

File: technology/kotlin-quarkus/src/test/kotlin/org/acme/kotlin/schooltimetabling/rest/TimetableResourceTest.kt
package org.acme.kotlin.schooltimetabling.rest

import ai.timefold.solver.core.api.solver.SolverStatus
import io.quarkus.test.junit.QuarkusTest
import io.restassured.RestAssured.get
import io.restassured.RestAssured.given
import io.restassured.http.ContentType
import org.acme.kotlin.schooltimetabling.domain.Room
import org.acme.kotlin.schooltimetabling.domain.Timeslot
import org.acme.kotlin.schooltimetabling.domain.Timetable
import org.awaitility.Awaitility.await
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertNotNull
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test
import java.time.Duration

@QuarkusTest
class TimetableResourceTest {

    @Test
    fun solveDemoDataUntilFeasible() {
        val testTimetable: Timetable = given()
            .`when`()["/demo-data/SMALL"]
            .then()
            .statusCode(200)
            .extract()
            .`as`(Timetable::class.java)

        val jobId: String = given()
            .contentType(ContentType.JSON)
            .body(testTimetable)
            .expect().contentType(ContentType.TEXT)
            .`when`().post("/timetables")
            .then()
            .statusCode(200)
            .extract()
            .asString()

        await()
            .atMost(Duration.ofMinutes(1))
            .pollInterval(Duration.ofMillis(500L))
            .until {
                SolverStatus.NOT_SOLVING.name ==
                        get("/timetables/$jobId/status")
                            .jsonPath().get("solverStatus")
            }
        val solution: Timetable =
            get("/timetables/$jobId").then().extract().`as`<Timetable>(
                Timetable::class.java
            )
        assertEquals(solution.solverStatus, SolverStatus.NOT_SOLVING)
        assertNotNull(solution.lessons)
        assertNotNull(solution.timeslots)
        assertNotNull(solution.rooms)
        assertNotNull(solution.lessons.get(0).room)
        assertNotNull(solution.lessons.get(0).timeslot)
        assertTrue(solution.score?.isFeasible!!)
    }

    @Test
    fun analyze() {
        val testTimetable: Timetable = given()
            .`when`()["/demo-data/SMALL"]
            .then()
            .statusCode(200)
            .extract()
            .`as`(Timetable::class.java)

        val roomList: List<Room> = testTimetable.rooms
        val timeslotList: List<Timeslot> = testTimetable.timeslots
        var i = 0
        for (lesson in testTimetable.lessons) { // Initialize the solution.
            lesson.room = roomList[i % roomList.size]
            lesson.timeslot = timeslotList[i % timeslotList.size]
            i += 1
        }
        val analysis: String = given()
            .contentType(ContentType.JSON)
            .body(testTimetable)
            .expect().contentType(ContentType.JSON)
            .`when`()
            .put("/timetables/analyze")
            .then()
            .extract()
            .asString()
        assertNotNull(analysis) // Too long to validate in its entirety.

        val analysis2: String = given()
            .contentType(ContentType.JSON)
            .queryParam("fetchPolicy", "FETCH_SHALLOW")
            .body(testTimetable)
            .expect().contentType(ContentType.JSON)
            .`when`()
            .put("/timetables/analyze")
            .then()
            .extract()
            .asString()
        assertNotNull(analysis2) // Too long to validate in its entirety.
    }
}
